learn/generation/langchain/rag-chatbot.ipynb

```diff
@@ -1,1163 +1,2742 @@
+        }
+      ],
+      "source": [
+        "dataset[0]"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "CKuloLmWZrwA"
+      },
+      "source": [
+        "#### Dataset Overview\n",
+        "\n",
+        "The dataset we are using is sourced from the Deepseek R1 ArXiv papers. Each entry in the dataset represents a \"chunk\" of text from the R1 paper.\n",
+        "\n",
+        "Because most **L**arge **L**anguage **M**odels (LLMs) only contain knowledge of the world as it was during training, even many of the newest LLMs cannot answer questions about Deepseek R1 — at least not without this data."
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "8KigzfQ2ZrwA"
+      },
+      "source": [
+        "### Task 4: Building the Knowledge Base"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "vtUCoidMZrwA"
+      },
+      "source": [
+        "We now have a dataset that can serve as our chatbot knowledge base. Our next task is to transform that dataset into the knowledge base that our chatbot can use. To do this we must use an embedding model and vector database.\n",
+        "\n",
+        "We begin by initializing our connection to Pinecone, this requires a [free API key](https://app.pinecone.io)."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 15,
+      "metadata": {
+        "id": "yg65GnRWZrwL"
+      },
+      "outputs": [],
+      "source": [
+        "from pinecone import Pinecone\n",
+        "\n",
+        "# initialize connection to pinecone (get API key at app.pinecone.io)\n",
+        "api_key = os.getenv(\"PINECONE_API_KEY\") or getpass(\n",
+        "    \"Enter your Pinecone API key: \"\n",
+        ")\n",
+        "\n",
+        "# configure client\n",
+        "pc = Pinecone(api_key=api_key)"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "WN6x5Y3vZrwL"
+      },
+      "source": [
+        "Now we setup our index specification, this allows us to define the cloud provider and region where we want to deploy our index. You can find a list of all [available providers and regions here](https://docs.pinecone.io/troubleshooting/available-cloud-regions)."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 16,
+      "metadata": {
+        "id": "T1LKJchlZrwL"
+      },
+      "outputs": [],
+      "source": [
+        "from pinecone import ServerlessSpec\n",
+        "\n",
+        "spec = ServerlessSpec(\n",
+        "    cloud=\"aws\", region=\"us-east-1\"\n",
+        ")"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "CK13RD1jZrwL"
+      },
+      "source": [
+        "Then we initialize the index. We will be using OpenAI's `text-embedding-ada-002` model for creating the embeddings, so we set the `dimension` to `1536`."
```

Comment:
You need to update this text to refer to `ext-embedding-3-small`

Reply:
added